lib/services/local/resource.go

```diff
@@ -0,0 +1,365 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+)
+
+// CreateResources attempts to dynamically create the supplied resources.
+// This function returns `trace.AlreadyExistsError` if one or more resources
+// would be overwritten, and `trace.NotImplementedError` if any resources
+// are of an unsupported type (see `ItemizeResource(...)`).
+//
+// NOTE: This function is non-atomic and performs no internal synchronization;
+// backend must be locked by caller when operating in parallel environment.
+func CreateResources(b backend.Backend, resources ...services.Resource) error {
+	var items []*backend.Item
+	// itemize all resources & ensure that they do not exist.
+	for _, r := range resources {
+		item, err := ItemizeResource(r)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		_, err = b.Get(context.TODO(), item.Key)
+		if !trace.IsNotFound(err) {
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			return trace.AlreadyExists("resource %q already exists", string(item.Key))
+		}
+		items = append(items, item)
+	}
+	// create all items.
+	for _, item := range items {
+		_, err := b.Create(context.TODO(), *item)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// ItemizeResource attempts to construct an instance of `backend.Item` from
+// a given resource.  If `rsc` is not one of the supported resource types,
+// a `trace.NotImplementedError` is returned.
+func ItemizeResource(resource services.Resource) (*backend.Item, error) {
+	// If resource is of unknown type, attempt to cast it to a concrete
+	// type before continuing.
+	if u, ok := resource.(*services.UnknownResource); ok {
+		r, err := services.UnmarshalResource(u.GetKind(), u.Raw)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resource = r
+	}
+	var item *backend.Item
+	var err error
+	switch r := resource.(type) {
+	case services.User:
+		item, err = itemizeUser(r)
```

Comment:
can you use the registered marshalers that you added here instead of another type switch?

Reply:
I can, and I started to at one point, but then decided against it.  Basically, in order to attach the `Resource <=> Item` conversions to the marshalers, I'd need to make a decent number of organizational changes, including making a lot values which are currently private members of `lib/services/local` into public members of `lib/services`.

At the time, I decided this was counter to the prevailing organizational philosophy since the `lib/services/local` package keeps all the logic related to inserting resources into the backend heavily encapsulated.

What do you think is the right move here?